lib/backend/watchersyncer/watchercache.go

```diff
@@ -122,53 +130,73 @@ func (wc *watcherCache) resyncAndCreateWatcher() {
+				wc.logger.WithError(err).Info("Failed to perform list of current data during resync")
+				time.Sleep(ListRetryInterval)
+				continue
+			}
 
-		// Send updates for each of the resources we listed - this will revalidate entries in
-		// the oldResources map.
-		for _, kvp := range l.KVPairs {
-			wc.handleWatchListEvent(kvp)
-		}
+			// Move the current resources over to the oldResources
+			wc.oldResources = wc.resources
+			wc.resources = make(map[string]cacheEntry, 0)
+
+			// Send updates for each of the resources we listed - this will revalidate entries in
+			// the oldResources map.
+			for _, kvp := range l.KVPairs {
+				wc.handleWatchListEvent(kvp)
+			}
+
+			// We've listed the current settings.  Complete the sync by notifying the main WatcherSyncer
+			// go routine (if we haven't already) and by sending deletes for the old resources that were
+			// not acknowledged by the List.  The oldResources will be empty after this call.
+			wc.finishResync()
 
-		// We've listed the current settings.  Complete the sync by notifying the main WatcherSyncer
-		// go routine (if we haven't already) and by sending deletes for the old resources that were
-		// not acknowledged by the List.  The oldResources will be empty after this call.
-		wc.finishResync()
+			// Store the current watch revision.  This gets updated on any new add/modified event.
+			wc.currentWatchRevision = l.Revision
+		}
 
-		// And now start watching from the revision returned by the List.
-		w, err := wc.client.Watch(context.Background(), wc.resourceType.ListInterface, l.Revision)
+		// And now start watching from the revision returned by the List, or from a previous watch event
+		// (depending on whether we were performing a full resync).
+		w, err := wc.client.Watch(context.Background(), wc.resourceType.ListInterface, wc.currentWatchRevision)
 		if err != nil {
 			// Failed to create the watcher - we'll need to retry.  Sleep so that we don't
-			// tight loop.  Since we have just performed a list, we need a slightly longer
+			// tight loop.  If we have just performed a list, we need a slightly longer
 			// delay to avoid overloading the datastore.  If the watcher keeps failing then
 			// we are effectively operating in a polling mode, so the interval should be a
 			// sensible polling interval.  Some resource types cannot be watched, so receiving
 			// an error here is not necessarily an error condition.
-			wc.logger.WithError(err).Debug("Failed to create watcher")
-			time.Sleep(WatchPollInterval)
+			wc.logger.WithError(err).WithField("performFullResync", performFullResync).Debug("Failed to create watcher")
+			if performFullResync {
+				wc.logger.Debug("Wait for watch poll interval")
+				time.Sleep(WatchPollInterval)
+			} else {
+				wc.logger.Debug("Wait for list retry interval")
+				time.Sleep(ListRetryInterval)
+			}
+
+			// Failure to watch, trigger a full resync.
+			performFullResync = true
```

Comment:
Hmm, I'm not fully convinced by the logic and comments here.

1. If "some resource types cannot be watched", that is presumably something that we can predict for each resource type?  In that case - and given that it feels like <failure to watch a resource that we don't expect to be watchable> should be treated quite differently from <failure to watch a resource that *should* be watchable> - then it feels like canBeWatched should be a property of this object that is set by whatever toplevel it is that creates an object for each resource type.

2. Then, if this was a watcher for which canBeWatched is false, why would we even try to watch it at all?

3. In the specific code just above, why sometimes sleep for ListRetryInterval, instead of WatchPollInterval, when just about to do a full resync?

4. For resource types that can be watched, what are the conditions that can cause a Watch() call to fail?  Is doing a full resync the right recovery, for all of those conditions?

Reply:
1. Yup we certainly could add that additional flag and I had considered it but omitted it originally due to time contraints, and in the case of not being able to watch we'd get an OperationNotSupported error back from KDD (since all etcdv3 resources are watchable) without any round trips to the API server.  So yes something I can add if you think it's necessary at this stage.

2. Yup we could do, just more stuff to plumb through and didn't seem immediately necessary since the cost of getting an operation not supported should be pretty small since it won't require any round trips to etcd. 

3.  The Watch poll is longer than the List retry interval - I just wanted to minimize the delay when doing an initial re-watch that wasn't actually possible because (for example) of compaction (the revision is no longer available) - in this case we shouldn't need to wait that long - in actual fact in this scenario I don't think I should be pausing at all - so i think I should pause for the Watch interval if we were doing a full resend, and if not don't pause at all.  WDYT?

Reply:
4.  Yup I agree and I did start looking into the different error cases that we should handle by retrying the watch rather than retrying the full resync.  However - that felt like a scarier change to rush in to a release without decent amounts of testing.  The code change in this PR, I believe, reinstates the same effective behavior that you'd implemented in the old KDD code - that was certainly the intention, but if that's not true then that is an issue of itself.

Reply:
In terms of etcd error types, I think the only error type that requires a full resync is a 401 but I think I'd rather make a change like that post-Beta when we can do a bunch of failure tests and check recovery is sensible.

Reply:
OK, I did some code tracing through kubernetes and the API server.  I _think_ I've found the source of the 1 minute timeout:

https://github.com/kubernetes/apiextensions-apiserver/blob/master/pkg/apiserver/customresource_handler.go#L187

Which then gets used the API server rest handler:

https://github.com/kubernetes/apiserver/blob/release-1.8/pkg/endpoints/handlers/rest.go#L359

From looking at this code, I think it should be sufficient to ensure we add a sensible TimeoutSeconds on the ListOptions that we use for CRDs, so I will add that too as part of this PR.